### PR TITLE
Make the agency sign-up select dropdown searchable

### PR DESCRIPTION
**Pull request by wilmersondasilva**
This select dropdown did not allow the users to type ahead. 

#### Proposed Changes

* This PR introduces a solution for this using the react-select library.

#### Testing Instructions

* Go to agency/signup
* Use the Country select, and try to search for some country
* Select some country that has available states, like the United States or Brazil, and then try to use the State select to search for some state.

#### Screenshots

image

image

image

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to #

### CODE REVIEW

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx
Comment (kallehauge) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: I'm not sure if the "disabled/loading" state is working as expecting. Personally I would expect the field to be greyed out or something similar while it's being loaded but we get a `- Select Country -` value in the field instead like the field is empty:

Screenshot 2022-10-07 at 15 34 08

And then, when it's done loading, the fields with the existing values are shown:

Screenshot 2022-10-07 at 15 34 15

File: client/jetpack-cloud/sections/partner-portal/company-details-form/components/searchable-dropdown.tsx
Comment (kallehauge) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/components/searchable-dropdown.tsx]: The value of the select field seems to be displayed in a bolder weight (at least with my Chrome v. 105.0.5195.125):

Screenshot 2022-10-07 at 15 41 40

**Review by kallehauge — CHANGES_REQUESTED** [review]
Awesome first take!

I've got a couple of lights comments and it seems like we get an error in our `Unit tests (Web app)` check saying: `Process exited with code 1 (Step: Prevent duplicated packages)`. Perhaps the `react-select` library already exists?

File: client/jetpack-cloud/sections/partner-portal/company-details-form/components/searchable-dropdown.tsx
Comment (wilmersondasilva) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/components/searchable-dropdown.tsx]: @kallehauge, I purposely let it as it was before. I didn't want to change it since I presumed it was approved by someone in the past. This is a screenshot of how it looks in the `trunk` branch.

image

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx
Comment (wilmersondasilva) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx:166]: You're right. The reason is the `countryOptions.length` will never be 0, because it'll always have at least one option, the `- Select Country -` option. I let it this way because it was what we had before. I just fixed it, and now the select input will be disabled while it loads the list of countries and will also show a loading indicator on its right.

image

Comment (wilmersondasilva) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: You're right. The reason for this is the `countryOptions.length` will never 0, since it's has always the option `- Select Country -`. I just fixed the code to deal with this.

image

This is what the users will see while the countries are fetched.

image

This was the behavior expected, but it wasn't working because of the bug I found. It's relevant to notice that the list of countries is fetched only once and then cached.

**Comment by robertf4** [comment]
I found a potential in-house alternative to `react-select` made by VIP. I asked [here](pdOhbm-bY-p2#comment-83) if we can use it in Calypso. If so then that might be preferable to pulling in a new third party library, and it seems that VIP has handled A11Y better than `react-select`.

**Comment by djalmaaraujo** [comment]
> I found a potential in-house alternative to `react-select` made by VIP. I asked [here](pdOhbm-bY-p2#comment-83) if we can use it in Calypso. If so then that might be preferable to pulling in a new third party library, and it seems that VIP has handled A11Y better than `react-select`.

@robertf4 I replied to your question. Another alternative if you don't want to mix up the libraries because of the dependencies is using directly the https://alphagov.github.io/accessible-autocomplete/ component we used as the base component.

This is probably the best component for autocomplete regarding accessibility and usability.

If you don't want to include the vip-design-system library, you can look at our implementation https://github.com/Automattic/vip-design-system/blob/trunk/src/system/NewForm/FormAutocomplete.js.

There's also something important to mention. At VIP we decided to separate displaying information with up to 15 options, and displaying information with more than 15 options.

For less than 15, we use the https://vip-design-system-components.netlify.app/?path=/story/form-select--default

For more, we use the Autocomplete solution: https://vip-design-system-components.netlify.app/?path=/story/form-autocomplete--default

This is to keep things accessible. There's no `Select` solution in the market that is more accessible than the regular `<select>` component.

Happy to help!

**Comment by DustinHartzler** [comment]
I tried testing this and got an error message loading the `/agency/signup/` page:


It looks like the `react-select` library isn't loading properly.

**Comment by DustinHartzler** [comment]
I needed to run `yarn install` first, whoops!

Works as expected. 

One improvement I would suggest is to clear out the field when it becomes focused, so it's more apparent you can type to search.

CleanShot 2022-11-30 at 11 53 46

**Comment by DustinHartzler** [comment]
Updated the placeholder text of the country field as well as matching the placeholder colors of the Business Address field. 

image


## Testing Instructions

- Go to http://jetpack.cloud.localhost:3000/agency/signup in an incognito window
- You'll need to create a new WordPress.com test account or use a test account that has not signed up for an agency before
- Use the Country select, and try to search for some country
- Select some country that has available states, like the United States or Brazil, and then try to use the State select to search for some state.

File: client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx
Comment (yashwin) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx]: How about we replace this with the [find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) method?

**Review by yashwin — CHANGES_REQUESTED** [review]
I had a look at this PR, and it works as expected. 

I left some comments and found a few UI/UX issues.

1) The border and highlight colors used here is blue in color, which I think should be black in color like other input boxes.

Screenshot 2022-12-12 at 6 40 27 PM

**Other input** 

Screenshot 2022-12-12 at 6 55 25 PM

2) When searching for an item in the country field, the text `Type to find country` is considered a dropdown item that also shows/hides based on the search term.

**Visible**

Screenshot 2022-12-12 at 6 38 13 PM

**Hidden**

Screenshot 2022-12-12 at 6 38 25 PM


Can we also rebase this PR with the latest trunk?

File: client/jetpack-cloud/sections/partner-portal/company-details-form/components/searchable-dropdown.tsx
Comment (tyxla) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/components/searchable-dropdown.tsx]: I'd discourage introducing yet another third-party component for this, given that we have our own alternatives.

Unless I'm missing something, we should be able to use `ComboboxControl` from `@wordpress/components` - see [component docs](https://developer.wordpress.org/block-editor/reference-guides/components/combobox-control/) for more info.

This component is already in use in a couple of places of Calypso:

https://github.com/Automattic/wp-calypso/blob/349a6b6a0fe384efcc8651ce47f0db4f4bf46fa1/client/landing/stepper/declarative-flow/internals/steps-repository/store-address/index.tsx#L246

https://github.com/Automattic/wp-calypso/blob/349a6b6a0fe384efcc8651ce47f0db4f4bf46fa1/client/signup/steps/woocommerce-install/step-store-address/index.tsx#L164

So I don't see a good reason to use another library for the same purpose. If there's a feature missing, we still might prefer to contribute upstream to support that feature, instead of using another 3rd-party library.

Here's some context as to why `@wordpress/components` are preferred:

p3N5Fy-uu-p2

p58i-7Ei-p2

Hope that helps, but please let me know if you run into any trouble or if you have any further questions.

**Review by tyxla — CHANGES_REQUESTED** [review]

File: client/jetpack-cloud/sections/partner-portal/company-details-form/components/searchable-dropdown.tsx
Comment (jkguidaven) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/components/searchable-dropdown.tsx]: @tyxla I tried to use the `ComboboxControl`, but it doesn't seem to support the `disabled` props. In case we want to contribute by adding the missing feature. How does that work on the timing of release? Do we immediately get the changes in Calypso, or will we have to wait for the next Gutenberg release?

Comment (jkguidaven) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/components/searchable-dropdown.tsx]: Ohh I just found out. It looks like there is a weekly release.

Comment (tyxla) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/components/searchable-dropdown.tsx]: I also wonder if wrapping the `ComboboxControl` in a `<Disabled />` component will do the job for you.

Comment (jkguidaven) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/components/searchable-dropdown.tsx]: @tyxla that works. 👍 I'll use that instead. Thanks.



I was looking at `inert`, but react doesn't play nice with it and makes the implementation hacky.

```
ref={ ( node ) => node && ( disabled ? node.setAttribute( 'inert', '' ) : node.removeAttribute( 'inert' ) )
```

Comment (tyxla) [reply client/jetpack-cloud/sections/partner-portal/company-details-form/components/searchable-dropdown.tsx]: Glad to hear it worked for you 👍

File: client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss
Comment (jkguidaven) [inline client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss:4]: You may wonder why this is so. Currently, an inline class is injected into the inner component with styling that always prevails.

Instead of using `!important` , this is another approach to force override.

**Comment by jkguidaven** [comment]
Updated the PR to use the suggested combobox from `@wordpress/components` package.

File: client/jetpack-cloud/sections/partner-portal/searchable-dropdown/index.tsx
Comment (tyxla) [inline client/jetpack-cloud/sections/partner-portal/searchable-dropdown/index.tsx]: If `Disabled` supports a `isDisabled` prop that you can use by passing `disabled` to it. That way you won't have to render two separate `ComboboxControl` instances.

File: client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss
Comment (tyxla) [inline client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss]: This should be coming by default from the `Disabled` component.

File: client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss
Comment (tyxla) [inline client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss:19]: It's a real bummer we have to override so many styles. 😞 

I wonder if some of them can be centrally applied to properly adapt and style the combobox control better for the entirety of Calypso 🤔

File: client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts
Comment (tyxla) [inline client/jetpack-cloud/sections/partner-portal/company-details-form/hooks/use-countries-and-states.ts:117]: 👍

**Review by tyxla — COMMENTED** [review]
Thanks for adapting it to use `@wordpress/components` 👍 

I've left some suggestions on the code and components side of things. Let me know what you think

File: client/jetpack-cloud/sections/partner-portal/searchable-dropdown/index.tsx
Comment (jkguidaven) [reply client/jetpack-cloud/sections/partner-portal/searchable-dropdown/index.tsx]: Though this work but the type definition didn't include the `isDisabled` props, which makes CI complain that it doesn't exist, 🤔 I'm not sure if there is something we can do about it.

```
import { ComponentType, Consumer, ReactNode } from 'react';

declare namespace Disabled {
    interface Props {
        children: ReactNode;
    }
}
declare const Disabled: {
    (props: Disabled.Props): JSX.Element;
    Consumer: Consumer<boolean>;
};

export default Disabled;
```

File: client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss
Comment (jkguidaven) [reply client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss]: Good catch 👍 this has been fixed.

File: client/jetpack-cloud/sections/partner-portal/searchable-dropdown/index.tsx
Comment (tyxla) [reply client/jetpack-cloud/sections/partner-portal/searchable-dropdown/index.tsx]: Well, the type definition includes `isDisabled` in trunk: https://github.com/WordPress/gutenberg/blob/b439e5f1a305f1c49d0817772aa0e7b02fa6c863/packages/components/src/disabled/types.ts#L7

It was added 9 months ago in #42708. 

The `@wordpress/components` package now ships its own types, so when we upgrade to the new version of `@wordpress/components`, it will be able to benefit from those types. We're currently working on the upgrade in #77121 and we're hoping to conclude it in the following weeks.

In the meantime, feel free to add a temporary type patch or even a `@ts-ignore` and a `TODO` comment there, pointing out that upgrading the `@wordpress/components` package to use its own types will fix it.

Comment (jkguidaven) [reply client/jetpack-cloud/sections/partner-portal/searchable-dropdown/index.tsx]: Thanks for looking into this. Good to hear this will be fixed soon. I added the `ts-ignore` as suggested. 👍

File: client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss
Comment (jkguidaven) [reply client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss:19]: I agree with this. I would say we can maybe do this on a separate PR. A couple of areas in which the combo box is being used have similar styling. We can probably decide what would be the standard Calypso styling from there. wdyt?

Comment (tyxla) [reply client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss:19]: Sounds good, no need to address it here. Perhaps we can add a TODO comment then to remember to address it.

Comment (jkguidaven) [reply client/jetpack-cloud/sections/partner-portal/searchable-dropdown/style.scss:19]: This has been added. 👍

**Review by TheSteveK — APPROVED** [review]
It looks like the code has been well and truly picked through and refined.  Tested and LGTM so 🚢

**Review by yashwin — APPROVED** [review]
I have tested the changes, and it LGTM. Good job on this, @jkguidaven! 🚀

**Review by tyxla — APPROVED** [review]
Nothing further from me, either, feel free to go ahead 🚀

**Comment by jkguidaven** [comment]
@yashwin @TheSteveK @tyxla thank you for reviewing this PR 🙇

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8026009

Thank you @wilmersondasilva for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8026009).

**Pull request merged**
